packages/@react-aria/menu/src/useMenuItem.ts

```diff
@@ -188,14 +203,68 @@ export function useMenuItem<T>(props: AriaMenuItemProps, state: TreeState<T>, re
       if (!isFocusVisible()) {
         state.selectionManager.setFocused(true);
         state.selectionManager.setFocusedKey(key);
+        // focus immediately so that a focus scope opened on hover has the correct restore node
+        let isFocused = key === state.selectionManager.focusedKey;
+        if (isFocused && state.selectionManager.isFocused && document.activeElement !== ref.current) {
+          if (state.expandedKeys.size > 0 && !state.expandedKeys.has(key)) {
+            for (let expandedKey of state.expandedKeys) {
+              state.toggleKey(expandedKey);
+            }
+          }
+          focusSafely(ref.current);
+        }
+      }
+    },
+    onHoverChange: isHovered => {
+      if (isHovered && isMenuDialogTrigger) {
+        if (!openTimeout.current) {
+          openTimeout.current = setTimeout(() => {
+            onSubmenuOpen();
+          }, 200);
+        }
+      } else if (!isHovered) {
+        cancelOpenTimeout();
+      }
+    }
+  });
+
+  let {keyboardProps} = useKeyboard({
+    onKeyDown: (e) => {
+      // Ignore repeating events, which may have started on the menu trigger before moving
+      // focus to the menu item. We want to wait for a second complete key press sequence.
+      if (e.repeat) {
+        e.continuePropagation();
+        return;
+      }
+
+      switch (e.key) {
+        case ' ':
+          if (!isDisabled && state.selectionManager.selectionMode === 'none' && !isMenuDialogTrigger && closeOnSelect !== false && onClose) {
+            onClose();
+          }
+          break;
+        case 'Enter':
+          // The Enter key should always close on select, except if overridden.
+          if (!isDisabled && closeOnSelect !== false && !isMenuDialogTrigger && onClose) {
+            onClose();
+          }
+          break;
+        case 'ArrowRight':
```

Comment:
Should this change for RTL?

Reply:
> Should this change for RTL?

I think it's always RightArrow because it has to do with advancement, not physical placement. For example, if the menu is all the way on the right side of the screen, the sub dialog or sub menu might flip to open on the left. https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/ though I'll run it by intl team

> I'm not able to keyboard navigate to the link inside.